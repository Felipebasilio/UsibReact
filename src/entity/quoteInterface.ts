export interface QuotesInterface {
  id?: string;
  product: string;
  companyName: string;
  quoteDate: string;
  adress: string;
  quantity: string;
  email: string;
}