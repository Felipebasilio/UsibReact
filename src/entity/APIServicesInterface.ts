import { QuotesInterface } from './quoteInterface';

export interface fetchReturn {
    dataFound: QuotesInterface[]; 
    dataFromId: QuotesInterface;
    errorFound: { 
      errorExists: boolean, 
      errorMessage: string 
    };
  }