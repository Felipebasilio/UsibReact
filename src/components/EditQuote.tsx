import React, { useState } from 'react';
import { useLocation } from 'react-router-dom';
import { QuotesInterface } from '../entity/quoteInterface';
import { fetchData, putData } from '../API/APIServices';


const EditQuote = () => {

  var state: QuotesInterface = useLocation().state as QuotesInterface;

  var id = new URLSearchParams(useLocation().search).get('id');

  var quoteToEdit: QuotesInterface = {} as QuotesInterface;

  if (state.id) {
    quoteToEdit = state;
  } else if(id) {
    fetchData(String(id)).then((response) => {
      console.log(response.dataFromId)
      quoteToEdit = response.dataFromId 
    })
  } else {

      alert("ID inexistente")

  }

  const [quote, setquote] = useState<QuotesInterface>(quoteToEdit);

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    putData(quote)
    
  }; 
  
  const onChange = (event: React.FormEvent<HTMLInputElement> | React.FormEvent<HTMLSelectElement>): void => {
    const fieldValue = event.currentTarget.value; 
    setquote({...quote, [(event.target as Element).id]: fieldValue});    
    
  }
  
  return (
    <div className="container my-5">
      <div className="my-5">
        <h2 className="col-12">Alteração de dados do orçamento { quote.id }</h2>
      </div>
      
      <form className="row g-3" id="edit-quote-form" onSubmit={handleSubmit}>

      <div className="col-md-6 mt-2">
        <label htmlFor="companyName" className="form-label">Nome da empresa:</label>
        <input 
          type="text"
          className="form-control"
          id="companyName"
          data-testid={`companyName${quote.id}`}
          value={ quote.companyName }
          onChange={onChange} 
          required/>
        <div className="invalid-feedback">
          Digite um nome de empresa válido.
        </div>
      </div>

      <div className="col-md-6 mt-2">
        <label htmlFor="email" className="form-label">Email de contato:</label>
        <input 
          type="email" 
          className="form-control" 
          id="email"
          data-testid={`email${quote.id}`}
          value={ quote.email } 
          onChange={onChange} 
          required/>
        <div className="invalid-feedback">
          Digite um email válido.
        </div>
      </div>

      <div className="col-6 mt-2">
        <label htmlFor="adress" className="form-label">Endereço:</label>
        <input 
          type="text" 
          className="form-control" 
          id="adress"
          data-testid={`adress${quote.id}`}
          value={ quote.adress }
          onChange={onChange} 
          required/>
        <div className="invalid-feedback">
          Digite um endereço válido.
        </div>
      </div>

      <div className="col-6 mt-2">
        <label htmlFor="quote-date" className="form-label">Data do orçamento:</label>
        <input 
          type="date" 
          className="form-control" 
          id="quoteDate" 
          data-testid={`quoteDate${quote.id}`}
          value={ quote.quoteDate } 
          onChange={onChange} 
          required/>
        <div className="invalid-feedback">
          Selecione uma data válida.
        </div>
      </div>

      <div className="col-6 mt-2 ">
          <label htmlFor="product" className="form-label">Produto</label>
          <select 
            className="form-select 
            custom-select" 
            id="product"
            data-testid={`product${quote.id}`}
            value={ quote.product }
            onChange={onChange} 
            required>

            <option value={ quote.product } disabled>{ quote.product }</option>
            <option value="Parafuso">Parafuso</option>
            <option value="Broca">Broca</option>

          </select>
          <div className="invalid-tooltip">
            Please select a valid state.
          </div>
        </div>

      <div className="col-6 mt-2">
        <label htmlFor="quantity" className="form-label">Quantidade:</label>
        <input 
          type="number" 
          className="form-control" 
          id="quantity" 
          data-testid={`quantity${quote.id}`}
          value={ quote.quantity } 
          onChange={onChange} 
          required/>
        <div className="invalid-feedback">
          Digite uma quantidade(número) valida.
        </div>
      </div>

      <div className="col-12 d-flex justify-content-center my-5">
        <button className="btn btn-primary col-12" type="submit">Atualizar pedido de orçamento</button>
      </div>
      </form>
    </div>
  );
};

export default EditQuote;

