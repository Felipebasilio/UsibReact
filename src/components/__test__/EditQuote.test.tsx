import { QuotesInterface } from './../../entity/quoteInterface';
import ReactDOM from 'react-dom';
import { createMemoryHistory } from 'history'
import { render } from '@testing-library/react';
import { Router } from 'react-router-dom';
import EditQuote from '../EditQuote';
import renderer from 'react-test-renderer'



const props: QuotesInterface = {
    id: '1',
    product: 'Parafuso',
    companyName: 'Nome teste',
    quoteDate: '2020-01-01',
    adress: 'EndereÃ§o teste',
    quantity: '100',
    email: 'email@teste.com'
} 


jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
        pathname: `localhost:3000/editQuote?id=${props.id}`,
        state: props
    })
}));

it('renders EditQuote without crashing', () => {
    let div = document.createElement('div');    
    ReactDOM.render(<EditQuote />, div );
    ReactDOM.unmountComponentAtNode(div);    
})

it('renders EditQuote properly', () => {
    const history = createMemoryHistory();
    const { getByTestId } = render(
        <Router history={ history }>
            <EditQuote />
        </Router>
    );

    expect(getByTestId(`companyName${props.id}`)).toHaveValue(props.companyName)
    expect(getByTestId(`quoteDate${props.id}`)).toHaveValue(props.quoteDate)
    expect(getByTestId(`adress${props.id}`)).toHaveValue(props.adress)
    expect(getByTestId(`email${props.id}`)).toHaveValue(props.email)
    expect(getByTestId(`product${props.id}`)).toHaveValue(props.product)
    expect(getByTestId(`quantity${props.id}`)).toHaveValue(parseInt(props.quantity))
})

it('Matches snapshot', () => {
    const EditQuoteComponnent = renderer.create(<EditQuote />).toJSON();
    expect(EditQuoteComponnent).toMatchSnapshot();
});
