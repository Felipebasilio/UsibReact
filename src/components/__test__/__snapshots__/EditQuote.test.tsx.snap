// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<div
  className="container my-5"
>
  <div
    className="my-5"
  >
    <h2
      className="col-12"
    >
      Alteração de dados do orçamento 
      1
    </h2>
  </div>
  <form
    className="row g-3"
    id="edit-quote-form"
    onSubmit={[Function]}
  >
    <div
      className="col-md-6 mt-2"
    >
      <label
        className="form-label"
        htmlFor="companyName"
      >
        Nome da empresa:
      </label>
      <input
        className="form-control"
        data-testid="companyName1"
        id="companyName"
        onChange={[Function]}
        required={true}
        type="text"
        value="Nome teste"
      />
      <div
        className="invalid-feedback"
      >
        Digite um nome de empresa válido.
      </div>
    </div>
    <div
      className="col-md-6 mt-2"
    >
      <label
        className="form-label"
        htmlFor="email"
      >
        Email de contato:
      </label>
      <input
        className="form-control"
        data-testid="email1"
        id="email"
        onChange={[Function]}
        required={true}
        type="email"
        value="email@teste.com"
      />
      <div
        className="invalid-feedback"
      >
        Digite um email válido.
      </div>
    </div>
    <div
      className="col-6 mt-2"
    >
      <label
        className="form-label"
        htmlFor="adress"
      >
        Endereço:
      </label>
      <input
        className="form-control"
        data-testid="adress1"
        id="adress"
        onChange={[Function]}
        required={true}
        type="text"
        value="Endereço teste"
      />
      <div
        className="invalid-feedback"
      >
        Digite um endereço válido.
      </div>
    </div>
    <div
      className="col-6 mt-2"
    >
      <label
        className="form-label"
        htmlFor="quote-date"
      >
        Data do orçamento:
      </label>
      <input
        className="form-control"
        data-testid="quoteDate1"
        id="quoteDate"
        onChange={[Function]}
        required={true}
        type="date"
        value="2020-01-01"
      />
      <div
        className="invalid-feedback"
      >
        Selecione uma data válida.
      </div>
    </div>
    <div
      className="col-6 mt-2 "
    >
      <label
        className="form-label"
        htmlFor="product"
      >
        Produto
      </label>
      <select
        className="form-select  custom-select"
        data-testid="product1"
        id="product"
        onChange={[Function]}
        required={true}
        value="Parafuso"
      >
        <option
          disabled={true}
          value="Parafuso"
        >
          Parafuso
        </option>
        <option
          value="Parafuso"
        >
          Parafuso
        </option>
        <option
          value="Broca"
        >
          Broca
        </option>
      </select>
      <div
        className="invalid-tooltip"
      >
        Please select a valid state.
      </div>
    </div>
    <div
      className="col-6 mt-2"
    >
      <label
        className="form-label"
        htmlFor="quantity"
      >
        Quantidade:
      </label>
      <input
        className="form-control"
        data-testid="quantity1"
        id="quantity"
        onChange={[Function]}
        required={true}
        type="number"
        value="100"
      />
      <div
        className="invalid-feedback"
      >
        Digite uma quantidade(número) valida.
      </div>
    </div>
    <div
      className="col-12 d-flex justify-content-center my-5"
    >
      <button
        className="btn btn-primary col-12"
        type="submit"
      >
        Atualizar pedido de orçamento
      </button>
    </div>
  </form>
</div>
`;
